### PR TITLE
8 check unused macros

### PR DESCRIPTION
**Pull request by tak30**

### CODE REVIEW

File: lib/meandro/rules/unused_macros.ex [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    for {file, module_asts} <- files_and_asts,
+        {_module_name, ast} <- module_asts,
+        result <- analyze_module(file, ast, files_and_asts) do
+      result
+    end
+    |> List.flatten()
+  end
+
+  @impl Meandro.Rule
+  def is_ignored?({module, macro}, {module, macro}) do
```
Comment (elbrujohalcon) [inline lib/meandro/rules/unused_macros.ex:19]: You probably don't need `module` here. Just `macro` should be enough, but careful… can there be multiple versions of the same macro name, with different arguments? And if so… can we have a user that wants to ignore warnings on just one of them and not all?

File: lib/meandro/rules/unused_macros.ex [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    false
+  end
+
+  defp analyze_module(file, ast, files_and_asts) do
+    module_aliases = Meandro.Util.module_aliases(ast)
+    macros = macros(ast, module_aliases)
+
+    for macro_info <- macros do
+      unused = is_unused?(macro_info, files_and_asts)
+      macro = macro_info |> Map.get(:name)
```
Comment (elbrujohalcon) [inline lib/meandro/rules/unused_macros.ex:33]: Can we have multiple macros with the same name? Like…
```elixir
defmacro my_macro do
…
end
defmacro my_macro(x, second_param: y, thrid_param: z) do
…
end
```

File: test/rules/unused_macros/bad.exs [hunk 3]
```diff
@@ -0,0 +1,2 @@
+defmodule MeandroTest.MyMacroTest do
```
Comment (elbrujohalcon) [inline test/rules/unused_macros/bad.exs:1]: This doesn't look like a `bad` module to me. It seems more like a `none.exs`.

**Review by elbrujohalcon — CHANGES_REQUESTED** [review]
I think it's good, but it's missing support for repeated macro names with different arities…

Can we have multiple macros with the same name? Like…
```elixir
defmacro my_macro do
…
end
defmacro my_macro(x, second_param: y, thrid_param: z) do
…
end
```

I would like to see a test where only one of those two macro versions is used and Meandro emits a warning, but _only for the other one_.

File: lib/meandro/rules/unused_macros.ex (hunk 1 above)
Comment (tak30) [reply lib/meandro/rules/unused_macros.ex]: Fixed

File: lib/meandro/rules/unused_macros.ex (hunk 2 above)
Comment (tak30) [reply lib/meandro/rules/unused_macros.ex:33]: Yep, now it takes that into account

File: test/rules/unused_macros/bad.exs (hunk 3 above)
Comment (tak30) [reply test/rules/unused_macros/bad.exs]: That one is actually not used. Deleted

File: lib/meandro/rules/unused_macros.ex [hunk 4]
```diff
@@ -16,7 +16,7 @@ defmodule Meandro.Rule.UnusedMacros do
   end
 
   @impl Meandro.Rule
-  def is_ignored?({module, macro}, {module, macro}) do
+  def is_ignored?({macro, arity}, {macro, arity}) do
```
Comment (elbrujohalcon) [inline lib/meandro/rules/unused_macros.ex:19]: What about an extra pattern, like…
```elixir
def is_ignored?({macro, _arity}, macro), do: true
```
…to ignore all versions of a certain macro by name.

File: lib/meandro/rules/unused_macros.ex [hunk 5]
```diff
@@ -32,15 +32,15 @@ defmodule Meandro.Rule.UnusedMacros do
...
 
       if unused do
         %Meandro.Rule{
           file: file,
           rule: __MODULE__,
           line: line,
-          pattern: {module, macro},
-          text: "The macro #{macro} is unused"
+          pattern: {macro, arity},
+          text: "The macro #{macro} with arity #{arity} is unused"
```
Comment (elbrujohalcon) [inline lib/meandro/rules/unused_macros.ex:43]:
```suggestion
          text: "The macro #{macro}/#{arity} is unused"
```

**Review by elbrujohalcon — APPROVED** [review]
Some minor comments, but it's good overall.

File: lib/meandro/rules/unused_macros.ex [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+  defp collect_macros(other, {module_aliases, macros}) do
+    {other, {module_aliases, macros}}
+  end
+
+  defp arity(nil) do
+    0
+  end
+
+  defp arity(list) do
+    Kernel.length(list)
```
Comment (pbrudnick) [inline lib/meandro/rules/unused_macros.ex:135]: this could be just:
```suggestion
    length(list)
```

**Review by pbrudnick — CHANGES_REQUESTED** [review]
Minor comments:
- (subjective) tend to use `case` instead of `if`
- left another minor suggestion for length

Request to change:
- Merge main and adapt tests to the Helpers and clean aliases way

**Review by pablocostass — APPROVED** [review]

**Comment by tak30** [comment]
> Minor comments:
> 
> * (subjective) tend to use `case` instead of `if`
> * left another minor suggestion for length
> 
> Request to change:
> 
> * Merge main and adapt tests to the Helpers and clean aliases way

Changes made. Regarding case over if. I was actually writting everyting with cases but @pablocostass told me that in elixir they use ifs so I tried to use them :D

**Pull request merged**